Repo URL:

https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID:

5b26f2ee8fa25665af804acaf58ad4861adc49f2

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('science_data_large.csv')
print(data.head(15))
print(data.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X, y = data[['Temperature °C', 'Mols KCL']], data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=44)

## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model

sample_datapoint = X_test.head(1)
sample_y = y_test.head(1)
predicition = reg.predict(sample_datapoint)
print(f'Predicition: {predicition[0]}, Actual: {sample_y.to_numpy()[0]}')

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print(f'Score: {reg.score(X_test, y_test)}')
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print(f'Coefficients: {reg.coef_}')
print(f'Intercepts: {reg.intercept_}')

Predicition: 404603.7734033278, Actual: 376202.2571
Score: 0.8538502713524164
Coefficients: [ 879.95924243 1034.5340203 ]
Intercepts: -417285.5825814459


Report: The score is approximately 0.85385, which is a good but not perfect score. According to the SciKit-Learn documentation, the score represents the coefficient of determination, $R^2$. The best score would be 1.0, whereas the worst score would be 0. Mathematically, $R^2$ is the sum of the squared residuals divided by the sum of squared differences of each label from the label mean, subtracting that quotient from 1. So since the score is close to 1, but not extremely close, that means our predictions from our linear regression model are slightly off from the actual values.

Rounding to 5 significant digits: $h(x_1, x_2) = -417080 + 881.14x_1 + 1024.4x_2$. Where $x_1$ represents temperature (°C) and $x_2$ represents Mols KCL.

## Part 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
print(cross_val_score(LinearRegression(), X, y))
# Report on their finding and their significance

[0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]


Report: The return value of the `cross_val_score` function is array of scores of our estimator for each run of the cross validation according to the SciKit-Learn documentation. Cross validation essentially subdivides the training data into several sets (in this case 5) and will train it each time on 4 out of the 5 data sets and use the remaining one as a test set. Each score is the represents one instance of cross validation. All of the scores are roughly the same, around 0.85, which was similar to the score in the previous section. It is good, but still has room for improvement as it could be closer to 1.

## Part 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
X_new = PolynomialFeatures(2).fit_transform(X_train)
X_test_poly = PolynomialFeatures(2).fit_transform(X_test)
poly = LinearRegression().fit(X_new, y_train)
# Report on the metrics and output the resultant equation as you did in Part 3.
print(f'Score: {poly.score(X_test_poly, y_test)}')
print(f'Coefficients: {poly.coef_}')
print(f'Intercept: {poly.intercept_}')

Score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.17926628e-07 -2.19557705e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 1.551362220197916e-05


Report: The score for this model is exactly 1. This means that this polynomial regression model exactly fits the test data, meaning our predicitions are completely accurate. Realisitically, this might be too good to be true and might potentially suggest that we may have accidently overfit the data by not splitting it properly, however for this assignment, this is expected.

The resultant polynomial equation (rounded to 5 significant digits) is:
 $h(x_1, x_2) = 1.3727 \times 10^{-5} + 12x_1 - (1.2102 \times 10^{-7})x_2 + (3.0983 \times 10^{-11})x_1^2 + 2x_1x_2 + 0.028571x_2^2$